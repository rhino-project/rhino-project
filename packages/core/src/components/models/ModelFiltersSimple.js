import PropTypes from 'prop-types';

import { ModelFiltersProvider } from './ModelFiltersProvider';
import { useModelFiltersController } from '../../hooks/controllers';

export const ModelFiltersSimple = ({ children, ...props }) => {
  const controller = useModelFiltersController(props);

  return (
    <ModelFiltersProvider {...controller}>{children}</ModelFiltersProvider>
  );
};

ModelFiltersSimple.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ])
};
