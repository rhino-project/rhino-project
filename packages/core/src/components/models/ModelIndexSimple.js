import PropTypes from 'prop-types';

import ModelIndexProvider from './ModelIndexProvider';
import { useModelIndexController } from '../../hooks/controllers';
import { Spinner } from 'reactstrap';
import { useMemo } from 'react';

export const ModelIndexSimple = ({ children, fallback = false, ...props }) => {
  const controller = useModelIndexController({
    ...props
  });
  const { isInitialLoading } = controller;

  // Fallback mirrors React 18 Suspense
  const renderFallback = useMemo(() => {
    if (!isInitialLoading || !fallback) return children;

    if (fallback === true) return <Spinner />;

    return fallback;
  }, [children, fallback, isInitialLoading]);

  return (
    <ModelIndexProvider {...controller}>{renderFallback}</ModelIndexProvider>
  );
};

ModelIndexSimple.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ])
};

export default ModelIndexSimple;
