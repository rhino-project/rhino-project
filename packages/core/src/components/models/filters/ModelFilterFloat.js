import PropTypes from 'prop-types';
import { useFormContext, useWatch } from 'react-hook-form';

import { useEffect, useMemo } from 'react';
import { useModelFilterField, useFilterPill } from 'rhino/hooks/form';
import FilterFloat from 'rhino/components/forms/filters/FilterFloat';

const ModelFilterFloat = ({ model, path, ...props }) => {
  const { attribute, operatorPath } = useModelFilterField(model, path);
  const { resetField } = useFormContext();

  // FIXME: Exclusive min/max support
  const min = useMemo(() => attribute.minimum, [attribute]);
  const max = useMemo(() => attribute.maximum, [attribute]);

  const watch = useWatch({ name: operatorPath, defaultValue: '' });

  const { resetPill, setPill } = useFilterPill(operatorPath);

  // If the user clears the input, we want to reset the pill
  useEffect(() => {
    if (watch !== undefined && watch !== '') {
      setPill(watch);
    } else {
      resetPill();
      // FIXME: This is a hack to reset the field when clear all filters is clicked
      resetField(operatorPath);
    }
  }, [operatorPath, resetField, resetPill, setPill, watch]);

  return <FilterFloat path={operatorPath} min={min} max={max} {...props} />;
};

ModelFilterFloat.propTypes = {
  operator: PropTypes.string,
  path: PropTypes.string.isRequired
};

export default ModelFilterFloat;
