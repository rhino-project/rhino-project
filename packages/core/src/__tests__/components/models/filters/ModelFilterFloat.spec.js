import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import ModelFiltersSimple from '../../../../components/models/ModelFiltersSimple';
import ModelIndexSimple from '../../../../components/models/ModelIndexSimple';
import ModelFilterFloat from '../../../../components/models/filters/ModelFilterFloat';

describe('ModelFilterFloat', () => {
  const Wrapper = ({ children, ...props }) => {
    const queryClient = new QueryClient();

    return (
      <MemoryRouter {...props}>
        <QueryClientProvider client={queryClient}>
          <ModelIndexSimple model="blog">
            <ModelFiltersSimple>{children}</ModelFiltersSimple>
          </ModelIndexSimple>
        </QueryClientProvider>
      </MemoryRouter>
    );
  };

  it(`adds min as a prop`, () => {
    const { asFragment } = render(
      <ModelFilterFloat
        model={{ properties: { dummy: { type: 'integer', minimum: 5 } } }}
        path="dummy"
      />,
      {
        wrapper: Wrapper
      }
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it(`adds min as a prop with exclusiveMinimum`, () => {
    const { asFragment } = render(
      <ModelFilterFloat
        model={{
          properties: {
            dummy: { type: 'integer', minimum: 5, exclusiveMinimum: true }
          }
        }}
        path="dummy"
      />,
      {
        wrapper: Wrapper
      }
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it(`adds max as a prop`, () => {
    const { asFragment } = render(
      <ModelFilterFloat
        model={{ properties: { dummy: { type: 'integer', maximum: 5 } } }}
        path="dummy"
      />,
      {
        wrapper: Wrapper
      }
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it(`adds max as a prop with exclusiveMaximum`, () => {
    const { asFragment } = render(
      <ModelFilterFloat
        model={{
          properties: {
            dummy: { type: 'integer', maximum: 5, exclusiveMaximum: true }
          }
        }}
        path="dummy"
      />,
      {
        wrapper: Wrapper
      }
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
