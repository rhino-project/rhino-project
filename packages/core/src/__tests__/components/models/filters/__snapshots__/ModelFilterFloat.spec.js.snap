// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ModelFilterFloat > adds max as a prop 1`] = `
<DocumentFragment>
  <input
    aria-invalid="false"
    autocomplete="off"
    class="form-control"
    id="dummy"
    max="5"
    name="dummy"
    path="dummy"
    type="number"
  />
</DocumentFragment>
`;

exports[`ModelFilterFloat > adds max as a prop with exclusiveMaximum 1`] = `
<DocumentFragment>
  <input
    aria-invalid="false"
    autocomplete="off"
    class="form-control"
    id="dummy"
    max="4.999999999999999"
    name="dummy"
    path="dummy"
    type="number"
  />
</DocumentFragment>
`;

exports[`ModelFilterFloat > adds min as a prop 1`] = `
<DocumentFragment>
  <input
    aria-invalid="false"
    autocomplete="off"
    class="form-control"
    id="dummy"
    min="5"
    name="dummy"
    path="dummy"
    type="number"
  />
</DocumentFragment>
`;

exports[`ModelFilterFloat > adds min as a prop with exclusiveMinimum 1`] = `
<DocumentFragment>
  <input
    aria-invalid="false"
    autocomplete="off"
    class="form-control"
    id="dummy"
    min="5.000000000000001"
    name="dummy"
    path="dummy"
    type="number"
  />
</DocumentFragment>
`;
